k2/csrc/multisplit.h

```diff
@@ -0,0 +1,191 @@
+                                       const uint64_t *packed,
+                                       uint32_t *out_key, uint32_t *out_value);
+
+/* Map an element to a bucket using `bucket_mapping_func`.
+ *
+ *  bucket_of_element_i = bucket_mapping_func(elements[i])
+ *
+ * @param [in] elements         Pointer to the elements array.
+ * @param [in] num_elements     Number of elements in the `elements` and
+ *                              `mark` array.
+ * @param [in] bucket_mapping_func
+ *                              It is a function like object supporting the
+ *                              following call operator:
+ *                              `uint32_t operator(uint32_t i)`
+ *
+ * @param [out] mark            Pointer to an array contain bucket information.
+ *                              `mark[i]` indicates which bucket `elements[i]`
+ *                              belongs to.
+ */
+template <typename Lambda>
+static __global__ void MarkBinsGeneral(const uint32_t *elements,
+                                       int32_t num_elements,
+                                       Lambda bucket_mapping_func,
+                                       uint32_t *mark) {
+  int32_t my_id = threadIdx.x + blockIdx.x * blockDim.x;
+  if (my_id >= num_elements) return;
+  int32_t offset = blockDim.x * gridDim.x;
+
+  for (int32_t i = my_id; i < num_elements; i += offset) {
+    uint32_t my_val = elements[i];
+    uint32_t my_bucket = bucket_mapping_func(my_val);
+    mark[i] = my_bucket;
+  }
+}
+
+/* MultiSplit using RadixSort.
+ *
+ * @param [in] context      Context indicating where the input data pointers,
+ *                          e.g., `in_keys` and `in_values`, come from.
+ * @param [in] num_elements Number of elements in `in_keys`, `in_values`,
+ *                          `out_keys` and `out_values`.
+ * @param [in] bucket_mapping_func
+ *                          It is a function like object support the following
+ *                          call operator: `uint32_t operator(uint32_t i)`.
+ *                          If it is NULL, we assume it is an identity function.
+ *                          That is, `in_keys[i]` also represents the bucket
+ *                          that `in_keys[i]` belongs to.
+ * @param [in] in_keys      Pointer to the input keys array.
+ * @param [in] in_values    Pointer to the input values array.
+ * @param [in] out_keys     Pointer to the output keys array. If not NULL,
+ *                          it will contain a sorted version of `in_keys`. The
+ *                          sort is performed according to the bucket
+ *                          information of the input elements and it is sorted
+ *                          in ascending order.
+ * @param [in] out_values   Pointer to the output values array. If not NULL,
+ *                          it will contain a sorted version of `in_values`.
+ */
+template <typename Lambda>
+void MultiSplitKeyValuePairs(ContextPtr context, int32_t num_elements,
```

Comment:
With below comments, I would suggest we change the funcion name to SortPairs, no `bucket_mapping_func`.
In the case where in_keys are the true key, we just call `cub::SortPairs`.
In the case where in_keys are the mapped keys, user can just pass values as `Range(0, num_elements)` and get `keys_out` and `values_out` in another lambda, of course we can wrapp the two steps into a single function (something like `GetSortedIndex`)

Reply:
Or  they don't care about `keys_out` they can just pass into `values` here.

Reply:
The function name is from CUDPP and I think MultiSplit is a well known
algorithm and it helps explain the meaning of the arguments.

Reply:
With a bucket mapping function, it is more flexible without introducing much overhead, I think.

Reply:
I'm OK with the name even though we use Sort approach here instead of Historgram approach (as you said before it only supports bins < = 32), the point is, as I commented before, if `bucket_mapping_func==nullptr`, we can just call SortPairs ( I noticed you have updated this); If `bucket_mapping_func != nullptr`, we still don't need the pack-then-unpack approach, as I said here:

> In the case where in_keys are the true key, we just call cub::SortPairs.
In the case where in_keys are the mapped keys, user can just pass values as Range(0, num_elements) and get keys_out and values_out in another lambda, of course we can wrapp the two steps into a single function (something like GetSortedIndex)

We can save one kenerl call here and another advantage is it can support any value types. Another point is, the caller usually have generated the mapped_keys (as in `GetTransposeReorder`), we don't need map it again with `bucket_mapping_func`.

Reply:
I think MultiSplit should be a general interface. `GetTransposeReordering` is just a special case.
I just used `RadixSort` directly in `GetTransposeReordering` and now `MultiSplit` is not
used at all in k2. I am going to remove it.